design/design-doc-e2e-testing.md

```diff
@@ -0,0 +1,809 @@
+
+These non-goals are important but out of the scope of this proposal.
+
+## Proposal
+
+We would like to propose the adoption of “Specification by Example”
+([wikipedia], [book]) methodology by defining tests using [Cucumber] framework
+and its [godog] bindings (2k stars on GitHub) for managing Crossplane
+end-to-end/acceptance tests. [Cucumber wiki] quote: _“Cucumber is a software
+tool that supports behavior-driven development (BDD). Central to the Cucumber
+BDD approach is its ordinary language parser called [Gherkin]. It allows
+expected software behaviors to be specified in a logical language that customers
+can understand. As such, Cucumber allows the execution of feature documentation
+written in business-facing text”. _
+
+Cucumber is a mature project/approach used successfully for testing systems of
+various complexity and size, for example:
+
+* [Service Binding Operator]
+* [Conformance tests for servicebinding specification]
+* [Primaza]
+* [Elastic end-to-end testing]
+
+Gherkin language defines just a handful of keywords and its purpose is to
+improve the readability of feature files. Other than that, the test authors have
+a full flexibility to define a human oriented language that describes test
+actions.
+
+### Anatomy of Cucumber Feature
+
+* Consist of a number of scenarios
+* Each scenario consists of sequential number of steps
+    * Each step describes an interaction with the system as a blackbox, using
+      only tools and clients available to
+      users (e.g. kubectl)
+    * Each step is described with a human readable sentence, not exposing system
+      implementation details
+    * Steps grammar is defined by us, i.e. the language grows with the system
+      features
+    * Steps can be parametrized
+    * Steps are reusable across many scenarios/features
+* Support for table driven tests
+* Support for common fixture management in single place
+
+### Steps Implementations
+
+Cucumber has bindings for [all popular languages]. [Go] bindings enable native
+integration into [go testing]. The project is very mature and has a large
+community. In essence:
+
+* Each scenario step maps to a Go function, enabling us even to use additional
+  go libraries where appropriate
+* A common context can be used to share state between test steps (e.g. to
+  further simplify step naming)
+* Through a number of different hooks, arbitrary code could be injected into
+  test workflow, e.g. to create some common
+  fixtures or perform required cleanups without cluttering the test description
+* A number of extensions available (e.g. [kubedog], [godox])
+
+
+* The tests are executed using the standard Go approach → they could be easily
+  triggered/debugged from various IDEs as well.
+* The amount of Go code implementing steps is very low, and yet the implemented
+  steps are enough for declaring a large number of test cases
+
+### Declarative Testing
+
+Cucumber allows us to perform [both imperative and declarative testing].
```

Comment:
It might be worth adding some examples of output from these tests. Even if a test passes, it is useful to know number of retries, execution time for each step, etc.

Reply:
Sure, you can check it easily. On this branch you can execute the tests simply with

```
make build e2e
```

Reply:
> you can check it easily

Let's add an example to the document per @hasheddan's suggestion please. That way reviewers can get an idea of what we're dealing with without needing to check out the branch and run the code.